backends/dpdk/dpdkProgram.cpp

```diff
@@ -0,0 +1,282 @@
+  for (auto obj : prog->objects) {
+    if (auto s = obj->to<IR::Type_Struct>()) {
+      if (s->name.name == info->local_metadata_type) {
+        auto *annotations = new IR::Annotations(
+            {new IR::Annotation(IR::ID("__metadata__"), {})});
+        for (auto anno : s->annotations->annotations)
+          annotations->add(anno);
+        auto st =
+            new IR::DpdkStructType(s->srcInfo, s->name, annotations, s->fields);
+        structType.push_back(st);
+      } else if (args_struct_map->find(s->name.name) !=
+                 args_struct_map->end()) {
+        auto st = new IR::DpdkArgStructType(s->srcInfo, s->name, s->annotations,
+                                            s->fields);
+        structType.push_back(st);
+      } else if (s->name.name == info->header_type) {
+        auto *annotations = new IR::Annotations(
+            {new IR::Annotation(IR::ID("__packet_data__"), {})});
+        for (auto anno : s->annotations->annotations)
+          annotations->add(anno);
+        auto st = new IR::DpdkHeaderStructType(s->srcInfo, s->name,
+                                               s->annotations, s->fields);
+        structType.push_back(st);
+      }
+    }
+  }
+
+  for (auto kv : structure.metadata_types) {
+    auto s = kv.second;
+    auto st =
+        new IR::DpdkStructType(s->srcInfo, s->name, s->annotations, s->fields);
+    structType.push_back(st);
+  }
+
+  IR::IndexedVector<IR::DpdkAsmStatement> statements;
+  auto ingress_parser_converter =
+      new ConvertToDpdkParser(refmap, typemap, collector, csum_map);
+  auto egress_parser_converter =
+      new ConvertToDpdkParser(refmap, typemap, collector, csum_map);
+  for (auto kv : structure.parsers) {
+    if (kv.first == "ingress")
+      kv.second->apply(*ingress_parser_converter);
+    else if (kv.first == "egress")
+      kv.second->apply(*egress_parser_converter);
+    else
+      BUG("Unknown parser %s", kv.second->name);
+  }
+  auto ingress_converter =
+      new ConvertToDpdkControl(refmap, typemap, collector, csum_map);
+  auto egress_converter =
+      new ConvertToDpdkControl(refmap, typemap, collector, csum_map);
+  for (auto kv : structure.pipelines) {
+    if (kv.first == "ingress")
+      kv.second->apply(*ingress_converter);
+    else if (kv.first == "egress")
+      kv.second->apply(*egress_converter);
+    else
+      BUG("Unknown control block %s", kv.second->name);
+  }
+  auto ingress_deparser_converter =
+      new ConvertToDpdkControl(refmap, typemap, collector, csum_map);
+  auto egress_deparser_converter =
+      new ConvertToDpdkControl(refmap, typemap, collector, csum_map);
+  for (auto kv : structure.deparsers) {
+    if (kv.first == "ingress")
+      kv.second->apply(*ingress_deparser_converter);
+    else if (kv.first == "egress")
+      kv.second->apply(*egress_deparser_converter);
+    else
+      BUG("Unknown deparser block %s", kv.second->name);
+  }
+  auto s = createListStatement("ingress",
```

Comment:
is there multicast on dpdk?
shouldn't there be some more instructions in between ingress and egress?

Reply:
There is currently no support for multicast.